File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
vector<vector<int>>result;
public:
    void dfs(int n, int target, vector<int> &candidates, vector<int>a)
    {
        if(n==0)
            return;
        if(target==0)
        {result.push_back(a);
         return;}
   
        if(candidates[n-1]<=target)
        {
            a.push_back(candidates[n-1]);
            dfs(n,target-candidates[n-1],candidates,a);
            a.pop_back();
        }
        dfs(n-1,target,candidates,a);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        int n=candidates.size();
        vector<int>a;
        dfs(n,target,candidates,a);
        return result;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> resList=new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> sublist=new ArrayList<>();
        helper(candidates,target,sublist,0);
        return resList;
    }
    public void helper(int[] arr,int target,List<Integer> sublist,int start){
        if(target==0){
            resList.add(new ArrayList<>(sublist));
        }else if(target>0){
        for(int i=start;i<arr.length;i++){
            sublist.add(arr[i]);
            helper(arr,target-arr[i],sublist,i);
            sublist.remove(sublist.size()-1);
        }
        }
    }
}